### DISCUSSION TITLE
How to make the application restart only when a scenario fails ?

### DISCUSSION THREAD

**Discussion by zbrea** [getgauge/gauge#2103] [General]
I want to run all the scenarios for a feature spec for my Android mobile app (with gauge-java) without closing it on each scenario, using the __BeforeSpec__ hook.
The problem if a scenario fails, the application will get stuck on the error screen.
This screen does not have access to be able to disconnect or close the application.
This causes all remaining scenarios in the spec to fail (false bugs).

How to make the application restart only when a scenario fails ?

> I don't want to use __BeforeScenario__ because the android application takes about 12 seconds for each restart and 3 seconds for authentication so for 1000 test scenarios I lose 15000 seconds which makes more than 4 hours wasted.

![image](https://user-images.githubusercontent.com/59898499/132987410-23b61b36-33b6-4107-a911-af771cd704ff.png)
![image](https://user-images.githubusercontent.com/59898499/132985734-59461136-10e0-4da0-bb65-e4edd4a09032.png)

```
gauge -v
```
![image](https://user-images.githubusercontent.com/59898499/132987833-86ef5fb8-09b6-403f-836f-a6469bbe509d.png)

**Comment by sriv** [comment]
You can try using the `AfterScenario` hook, which takes in an [`ExecutionContext`](https://javadoc.io/doc/com.thoughtworks.gauge/gauge-java/latest/com/thoughtworks/gauge/Scenario.html) parameter that holds information about the current run:

```
    @AfterScenario
    public void afterScenarioHook(ExecutionContext ctx) {
        if (ctx.getCurrentScenario().getIsFailing()) {
            // restart app?
        }
    }
```

**Reply by zbrea** [reply]
This solution works wonderfully,
But what if the last scenario fails, so that the application does not restart if there is no more scenario to launch.
I want to add a condition that the next scenario (getCurrentScenario + 1) is not null

**Reply by sriv** [reply]
>  I want to add a condition that the next scenario (getCurrentScenario + 1) is not null

This unfortunately may not be possible today. You can perhaps move the restart logic to `BeforeScenario`, but then beware that the `currentScenario` will not be the one that failed, rather it will be the next scenario to be executed.

So, something like:

```
@BeforeScenario
public void beforeScenarioHook() {
\\\ if app is running
\\\ continue
\\\ else
\\\ restart app

}
```
